src/equations/serre_green_naghdi_flat_1d.jl

```diff
@@ -0,0 +1,313 @@
+        #
+        # Split form for energy conservation:
+        # h_t + h_x v + h v_x = 0
+        @. dh = -(h_x * v + h * v_x)
+
+        # Plain: h v_t + ... = 0
+        #
+        # Split form for energy conservation:
+        @. tmp = -(g * h2_x - g * h * h_x
+                   +
+                   0.5 * h * v2_x
+                   -
+                   0.5 * v^2 * h_x
+                   +
+                   0.5 * hv_x * v
+                   -
+                   0.5 * h * v * v_x
+                   +
+                   p_x)
+    end
+
+    @trixi_timeit timer() "assembling elliptic operator" begin
+        # The code below is equivalent to
+        #   dv .= (Diagonal(h) - Dmat * Diagonal(1/3 .* h.^3) * Dmat) \ tmp
+        # but faster since the symbolic factorization is reused.
+        # Floating point errors accumulate a bit and the system matrix
+        # is not necessarily perfectly symmetric but only up to round-off
+        # errors. We wrap it here to avoid issues with the factorization.
+        @. M_h = h
+        scale_by_mass_matrix!(M_h, D)
+        inv3 = 1 / 3
+        @. M_h3_3 = inv3 * h^3
+        scale_by_mass_matrix!(M_h3_3, D)
+        system_matrix = Symmetric(Diagonal(M_h)
+                                  +
+                                  Dmat' * Diagonal(M_h3_3) * Dmat)
+    end
+
+    @trixi_timeit timer() "solving elliptic system" begin
+        if issparse(system_matrix)
+            (; factorization) = cache
+            cholesky!(factorization, system_matrix; check = false)
+            if issuccess(factorization)
+                scale_by_mass_matrix!(tmp, D)
+                dv .= factorization \ tmp
+            else
+                # The factorization may fail if the time step is too large
+                # and h becomes negative.
+                fill!(dv, Inf)
+            end
+        else
+            factorization = cholesky!(system_matrix)
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+@inline function prim2cons(q, equations::SerreGreenNaghdiEquations1D)
+    h, v = q
+
+    hv = h * v
+    return SVector(h, hv)
+end
+
+@inline function cons2prim(u, equations::SerreGreenNaghdiEquations1D)
+    h, hv = u
+
+    v = hv / h
+    return SVector(h, v)
+end
+
+@inline function waterheight_total(q, equations::SerreGreenNaghdiEquations1D)
+    return waterheight(q, equations) + bathymetry(q, equations)
+end
+
+@inline function velocity(q, equations::SerreGreenNaghdiEquations1D)
+    return q[2]
+end
+
+@inline function bathymetry(q, equations::SerreGreenNaghdiEquations1D{BathymetryFlat})
+    return zero(eltype(q))
+end
+
+@inline function waterheight(q, equations::SerreGreenNaghdiEquations1D)
+    return q[1]
+end
+
+# The entropy/energy takes the whole `q` for every point in space
+@inline function energy_total(q, equations::SerreGreenNaghdiEquations1D{BathymetryFlat},
```

Comment:
As above, we could think of renaming this `energy_total_modified` and add another quantity `energy_total` computing only the pure shallow water energy part. Even if this is not conserved by the equations, it can be interesting how it evolves under the solution of the Serre-Green-Naghdi equations, I think.

Reply:
We decided to rename this to `energy_total_modified`. In this case, this function should get a proper docstring.

Reply:
Done in dca91a17927e0f4efedd47c698d23800e24bff9b